### ISSUE TITLE
git.latest attempting to update (and failing) when it's already up to date

### ISSUE THREAD

**Issue by boltronics** [saltstack/salt#36321]
### Description of Issue/Question
git.latest appears broken since upgrading to 2016.3.3.

My example state basically looks like this:
```
git@github.com:myaccount/myrepo.git:
  git.latest:
    - rev: {{ salt['pillar.get']('git:myrepo:rev') }}
    - target: /srv/www/myrepo
    - identity: /some_path/id_rsa.github
    - force_checkout: True
    - submodules: False

/srv/www/myrepo/somefile:
  file.append:
    - text: some string
    - require:
      - git: git@github.com:myaccount/myrepo.git
```

This worked great in 2016.3.2, but since upgrading it will error out with the following:
```
          ID: git@github.com:myaccount/myrepo.git
    Function: git.latest
      Result: False
     Comment: Repository would be updated to c15b1a6, but there are uncommitted changes. Set 'force_reset' to True to force this update and discard these changes.
     Started: 18:05:48.001508
    Duration: 25839.618 ms
     Changes:   
```

The problem? It's already on c15b1a6 so it doesn't need to force a checkout - it shouldn't be doing anything at all.

I can take the suggestion to add the ` - force_reset: True` argument, but then `/srv/www/myrepo/somefile` gets clobbered on every execution, causing both of these to report changes on every execution! Needless to say, this is not desirable and we may need to hold off upgrading. These states should only report changes when there is a requested update to the git codebase. Any known work-around to achieve this would be very much appreciated.

It looks like the issue might be related to https://github.com/saltstack/salt/issues/34725.

### Versions Report
```
# apt-cache policy salt-minion | head -n 3
salt-minion:
  Installed: 2016.3.3-1
  Candidate: 2016.3.3-1
# salt-call --versions-report
Salt Version:
           Salt: 2016.3.0
 
Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.2
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.7.3
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: 1.0.0
   msgpack-pure: Not Installed
 msgpack-python: 0.4.2
   mysql-python: 1.2.3
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: Not Installed
         Python: 2.7.9 (default, Mar  1 2015, 12:57:24)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 14.4.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.4.1
            ZMQ: 4.0.5
 
System Versions:
           dist: debian 8.5 
        machine: x86_64
        release: 3.16.0-4-amd64
         system: Linux
        version: debian 8.5 
 
# 
```

**Comment by boltronics** [comment]
As I understand it, force_reset's behaviour now does something quite different to what it used to do. Unless I'm overlooking something, this new functionality should probably have used a new argument so `git.latest` could work either way. As it stands, it broke backwards compatibility in a point release.

**Comment by boltronics** [comment]
It's not just my production deploys with the issue, as I have encountered another issue caused by this change. I have a development environment which has my production states cloned via git. I have a state that uses `git.latest` to make sure the master branch of my states are always up to date. The idea behind this development setup is that I like to test my changes are functioning as expected, and only then branch and commit my changes if everything looks good.

As you might imagine, now this setup won't work. Running highstate on this test salt master will result in the uncommitted changes error (as well as a bunch of failures for everything that depends on it), but I certainly don't want my uncommitted changes nuked!

With this additional functionality affected, now I know that there's no way I can use these new Salt versions without a fix or at least a work around.

**Comment by gtmanfred** [comment]
I believe that this is now the expected behavior, but I will need to check with the developer who wrote the code to change it.

**Comment by gtmanfred** [comment]
Are you making changes to a tracked file? If not, would it be possible to add that to the repos `.gitignore` so the git doesn't see the file at all when you make changes or add it to the repository?

**Comment by boltronics** [comment]
In most cases I'm making changes to a tracked file.

I think you are correct that the behaviour was deliberately changed, but I think it was a serious mistake to do so.

**Comment by terminalmage** [comment]
This is not the expected behavior. This was an attempt to address https://github.com/saltstack/salt/issues/34725 by allowing ``force_reset`` to reset local changes, but it went too far in that it identified local changes as a non-fast-forward change. It should not be doing that. I should have a fix for this in the next couple days.

**Comment by terminalmage** [comment]
OK, this should be resolved now in #36378.

## Using pre-fix code...

```
# salt-call --local state.single git.latest name=/tmp/test-repo target=/tmp/foo
15:29:42,142 [salt.state][ 766][INFO    ][16837] Loading fresh modules for state activity
15:29:42,171 [salt.loaded.int.module.cmdmod][ 360][INFO    ][16837] Executing command ['git', '--version'] in directory '/root'
15:29:42,176 [salt.state][1542][INFO    ][16837] Running state [/tmp/test-repo] at time 15:29:42.176391
15:29:42,177 [salt.state][1564][INFO    ][16837] Executing state git.latest for /tmp/test-repo
15:29:42,177 [salt.loaded.int.states.git][ 555][INFO    ][16837] Checking remote revision for /tmp/test-repo
15:29:42,177 [salt.loaded.int.module.cmdmod][ 360][INFO    ][16837] Executing command ['git', 'ls-remote', '/tmp/test-repo'] in directory '/root'
15:29:42,190 [salt.loaded.int.module.cmdmod][ 360][INFO    ][16837] Executing command ['git', 'rev-parse', '--show-toplevel'] in directory '/tmp/foo'
15:29:42,190 [salt.loaded.int.module.cmdmod][ 360][INFO    ][16837] Executing command ['git', 'clone', '--', '/tmp/test-repo', '/tmp/foo'] in directory '/tmp'
15:29:42,199 [salt.loaded.int.states.git][ 112][INFO    ][16837] Checking local revision for /tmp/foo
15:29:42,199 [salt.loaded.int.module.cmdmod][ 360][INFO    ][16837] Executing command ['git', 'rev-parse', 'HEAD'] in directory '/tmp/foo'
15:29:42,207 [salt.loaded.int.states.git][ 121][INFO    ][16837] Checking local branch for /tmp/foo
15:29:42,208 [salt.loaded.int.module.cmdmod][ 360][INFO    ][16837] Executing command ['git', 'rev-parse', '--abbrev-ref', 'HEAD'] in directory '/tmp/foo'
15:29:42,216 [salt.loaded.int.module.cmdmod][ 360][INFO    ][16837] Executing command ['git', 'rev-parse', '--abbrev-ref', 'master@{upstream}'] in directory '/tmp/foo'
15:29:42,222 [salt.loaded.int.module.cmdmod][ 360][INFO    ][16837] Executing command ['git', 'rev-parse', 'HEAD'] in directory '/tmp/foo'
15:29:42,226 [salt.loaded.int.states.git][1580][INFO    ][16837] /tmp/test-repo cloned to /tmp/foo
15:29:42,227 [salt.state][ 232][INFO    ][16837] {'new': '/tmp/test-repo => /tmp/foo', 'revision': {'new': '189ed248555abd3ce9f33318144034d640203773', 'old': None}}
15:29:42,227 [salt.state][1678][INFO    ][16837] Completed state [/tmp/test-repo] at time 15:29:42.227012
local:
----------
          ID: /tmp/test-repo
    Function: git.latest
      Result: True
     Comment: /tmp/test-repo cloned to /tmp/foo
     Started: 15:29:42.176391
    Duration: 50.621 ms
     Changes:
              ----------
              new:
                  /tmp/test-repo => /tmp/foo
              revision:
                  ----------
                  new:
                      189ed248555abd3ce9f33318144034d640203773
                  old:
                      None

Summary for local
------------
Succeeded: 1 (changed=1)
Failed:    0
------------
Total states run:     1
Total run time:  50.621 ms
# salt-call --local state.single git.latest name=/tmp/test-repo target=/tmp/foo
15:30:03,826 [salt.state][ 766][INFO    ][17050] Loading fresh modules for state activity
15:30:03,834 [salt.loaded.int.module.cmdmod][ 360][INFO    ][17050] Executing command ['git', '--version'] in directory '/root'
15:30:03,839 [salt.state][1542][INFO    ][17050] Running state [/tmp/test-repo] at time 15:30:03.838653
15:30:03,839 [salt.state][1564][INFO    ][17050] Executing state git.latest for /tmp/test-repo
15:30:03,839 [salt.loaded.int.states.git][ 555][INFO    ][17050] Checking remote revision for /tmp/test-repo
15:30:03,840 [salt.loaded.int.module.cmdmod][ 360][INFO    ][17050] Executing command ['git', 'ls-remote', '/tmp/test-repo'] in directory '/root'
15:30:03,847 [salt.loaded.int.module.cmdmod][ 360][INFO    ][17050] Executing command ['git', 'for-each-ref', '--format', '%(refname:short)', 'refs/heads/'] in directory '/tmp/foo'
15:30:03,852 [salt.loaded.int.module.cmdmod][ 360][INFO    ][17050] Executing command ['git', 'for-each-ref', '--format', '%(refname:short)', 'refs/tags/'] in directory '/tmp/foo'
15:30:03,857 [salt.loaded.int.states.git][ 112][INFO    ][17050] Checking local revision for /tmp/foo
15:30:03,857 [salt.loaded.int.module.cmdmod][ 360][INFO    ][17050] Executing command ['git', 'rev-parse', 'HEAD'] in directory '/tmp/foo'
15:30:03,862 [salt.loaded.int.states.git][ 121][INFO    ][17050] Checking local branch for /tmp/foo
15:30:03,862 [salt.loaded.int.module.cmdmod][ 360][INFO    ][17050] Executing command ['git', 'rev-parse', '--abbrev-ref', 'HEAD'] in directory '/tmp/foo'
15:30:03,869 [salt.loaded.int.module.cmdmod][ 360][INFO    ][17050] Executing command ['git', 'remote', '--verbose'] in directory '/tmp/foo'
15:30:03,874 [salt.loaded.int.module.cmdmod][ 360][INFO    ][17050] Executing command ['git', 'diff', 'HEAD'] in directory '/tmp/foo'
15:30:03,879 [salt.loaded.int.module.cmdmod][ 360][INFO    ][17050] Executing command ['git', 'merge-base', '--is-ancestor', '189ed248555abd3ce9f33318144034d640203773', '189ed248555abd3ce9f33318144034d640203773'] in directory '/tmp/foo'
15:30:03,885 [salt.loaded.int.module.cmdmod][ 360][INFO    ][17050] Executing command ['git', 'rev-parse', '--abbrev-ref', 'master@{upstream}'] in directory '/tmp/foo'
15:30:03,891 [salt.loaded.int.module.cmdmod][ 360][INFO    ][17050] Executing command ['git', 'rev-parse', 'HEAD'] in directory '/tmp/foo'
15:30:03,896 [salt.state][ 232][INFO    ][17050] Repository /tmp/foo is up-to-date
15:30:03,896 [salt.state][1678][INFO    ][17050] Completed state [/tmp/test-repo] at time 15:30:03.896435
local:
----------
          ID: /tmp/test-repo
    Function: git.latest
      Result: True
     Comment: Repository /tmp/foo is up-to-date
     Started: 15:30:03.838653
    Duration: 57.782 ms
     Changes:

Summary for local
------------
Succeeded: 1
Failed:    0
------------
Total states run:     1
Total run time:  57.782 ms
```

## After making local changes...

```
# echo "add some content" >>/tmp/foo/foo
# salt-call --local state.single git.latest name=/tmp/test-repo target=/tmp/foo
15:30:47,628 [salt.state][ 766][INFO    ][17423] Loading fresh modules for state activity
15:30:47,636 [salt.loaded.int.module.cmdmod][ 360][INFO    ][17423] Executing command ['git', '--version'] in directory '/root'
15:30:47,641 [salt.state][1542][INFO    ][17423] Running state [/tmp/test-repo] at time 15:30:47.640864
15:30:47,641 [salt.state][1564][INFO    ][17423] Executing state git.latest for /tmp/test-repo
15:30:47,642 [salt.loaded.int.states.git][ 555][INFO    ][17423] Checking remote revision for /tmp/test-repo
15:30:47,642 [salt.loaded.int.module.cmdmod][ 360][INFO    ][17423] Executing command ['git', 'ls-remote', '/tmp/test-repo'] in directory '/root'
15:30:47,649 [salt.loaded.int.module.cmdmod][ 360][INFO    ][17423] Executing command ['git', 'for-each-ref', '--format', '%(refname:short)', 'refs/heads/'] in directory '/tmp/foo'
15:30:47,655 [salt.loaded.int.module.cmdmod][ 360][INFO    ][17423] Executing command ['git', 'for-each-ref', '--format', '%(refname:short)', 'refs/tags/'] in directory '/tmp/foo'
15:30:47,659 [salt.loaded.int.states.git][ 112][INFO    ][17423] Checking local revision for /tmp/foo
15:30:47,660 [salt.loaded.int.module.cmdmod][ 360][INFO    ][17423] Executing command ['git', 'rev-parse', 'HEAD'] in directory '/tmp/foo'
15:30:47,667 [salt.loaded.int.states.git][ 121][INFO    ][17423] Checking local branch for /tmp/foo
15:30:47,667 [salt.loaded.int.module.cmdmod][ 360][INFO    ][17423] Executing command ['git', 'rev-parse', '--abbrev-ref', 'HEAD'] in directory '/tmp/foo'
15:30:47,676 [salt.loaded.int.module.cmdmod][ 360][INFO    ][17423] Executing command ['git', 'remote', '--verbose'] in directory '/tmp/foo'
15:30:47,684 [salt.loaded.int.module.cmdmod][ 360][INFO    ][17423] Executing command ['git', 'diff', 'HEAD'] in directory '/tmp/foo'
15:30:47,690 [salt.state][ 234][ERROR   ][17423] Repository would be updated to 189ed24, but there are uncommitted changes. Set 'force_reset' to True to force this update and discard these changes.
15:30:47,690 [salt.state][1678][INFO    ][17423] Completed state [/tmp/test-repo] at time 15:30:47.690199
local:
----------
          ID: /tmp/test-repo
    Function: git.latest
      Result: False
     Comment: Repository would be updated to 189ed24, but there are uncommitted changes. Set 'force_reset' to True to force this update and discard these changes.
     Started: 15:30:47.640864
    Duration: 49.335 ms
     Changes:

Summary for local
------------
Succeeded: 0
Failed:    1
------------
Total states run:     1
Total run time:  49.335 ms
```

## After the fix is implemented, we repeat the same command...

```
# salt-call --local state.single git.latest name=/tmp/test-repo target=/tmp/foo
15:31:33,465 [salt.state][ 766][INFO    ][17907] Loading fresh modules for state activity
15:31:33,485 [salt.loaded.int.module.cmdmod][ 360][INFO    ][17907] Executing command ['git', '--version'] in directory '/root'
15:31:33,490 [salt.state][1542][INFO    ][17907] Running state [/tmp/test-repo] at time 15:31:33.490439
15:31:33,491 [salt.state][1564][INFO    ][17907] Executing state git.latest for /tmp/test-repo
15:31:33,491 [salt.loaded.int.states.git][ 558][INFO    ][17907] Checking remote revision for /tmp/test-repo
15:31:33,491 [salt.loaded.int.module.cmdmod][ 360][INFO    ][17907] Executing command ['git', 'ls-remote', '/tmp/test-repo'] in directory '/root'
15:31:33,501 [salt.loaded.int.module.cmdmod][ 360][INFO    ][17907] Executing command ['git', 'for-each-ref', '--format', '%(refname:short)', 'refs/heads/'] in directory '/tmp/foo'
15:31:33,509 [salt.loaded.int.module.cmdmod][ 360][INFO    ][17907] Executing command ['git', 'for-each-ref', '--format', '%(refname:short)', 'refs/tags/'] in directory '/tmp/foo'
15:31:33,517 [salt.loaded.int.states.git][ 112][INFO    ][17907] Checking local revision for /tmp/foo
15:31:33,518 [salt.loaded.int.module.cmdmod][ 360][INFO    ][17907] Executing command ['git', 'rev-parse', 'HEAD'] in directory '/tmp/foo'
15:31:33,523 [salt.loaded.int.states.git][ 121][INFO    ][17907] Checking local branch for /tmp/foo
15:31:33,523 [salt.loaded.int.module.cmdmod][ 360][INFO    ][17907] Executing command ['git', 'rev-parse', '--abbrev-ref', 'HEAD'] in directory '/tmp/foo'
15:31:33,530 [salt.loaded.int.module.cmdmod][ 360][INFO    ][17907] Executing command ['git', 'remote', '--verbose'] in directory '/tmp/foo'
15:31:33,535 [salt.loaded.int.module.cmdmod][ 360][INFO    ][17907] Executing command ['git', 'diff', 'HEAD'] in directory '/tmp/foo'
15:31:33,543 [salt.state][ 232][INFO    ][17907] Repository /tmp/foo is up-to-date, but with local changes. Set 'force_reset' to True to purge local changes.
15:31:33,544 [salt.state][1678][INFO    ][17907] Completed state [/tmp/test-repo] at time 15:31:33.543741
local:
----------
          ID: /tmp/test-repo
    Function: git.latest
      Result: True
     Comment: Repository /tmp/foo is up-to-date, but with local changes. Set 'force_reset' to True to purge local changes.
     Started: 15:31:33.490439
    Duration: 53.302 ms
     Changes:

Summary for local
------------
Succeeded: 1
Failed:    0
------------
Total states run:     1
Total run time:  53.302 ms
```

## Using ``force_reset`` in test mode to show what *would* happen if we proceed

```
# salt-call --local state.single git.latest name=/tmp/test-repo target=/tmp/foo force_reset=True test=True
15:31:51,280 [salt.state][ 766][INFO    ][18125] Loading fresh modules for state activity
15:31:51,292 [salt.loaded.int.module.cmdmod][ 360][INFO    ][18125] Executing command ['git', '--version'] in directory '/root'
15:31:51,297 [salt.state][1542][INFO    ][18125] Running state [/tmp/test-repo] at time 15:31:51.297419
15:31:51,298 [salt.state][1564][INFO    ][18125] Executing state git.latest for /tmp/test-repo
15:31:51,298 [salt.loaded.int.states.git][ 558][INFO    ][18125] Checking remote revision for /tmp/test-repo
15:31:51,298 [salt.loaded.int.module.cmdmod][ 360][INFO    ][18125] Executing command ['git', 'ls-remote', '/tmp/test-repo'] in directory '/root'
15:31:51,306 [salt.loaded.int.module.cmdmod][ 360][INFO    ][18125] Executing command ['git', 'for-each-ref', '--format', '%(refname:short)', 'refs/heads/'] in directory '/tmp/foo'
15:31:51,311 [salt.loaded.int.module.cmdmod][ 360][INFO    ][18125] Executing command ['git', 'for-each-ref', '--format', '%(refname:short)', 'refs/tags/'] in directory '/tmp/foo'
15:31:51,317 [salt.loaded.int.states.git][ 112][INFO    ][18125] Checking local revision for /tmp/foo
15:31:51,318 [salt.loaded.int.module.cmdmod][ 360][INFO    ][18125] Executing command ['git', 'rev-parse', 'HEAD'] in directory '/tmp/foo'
15:31:51,326 [salt.loaded.int.states.git][ 121][INFO    ][18125] Checking local branch for /tmp/foo
15:31:51,327 [salt.loaded.int.module.cmdmod][ 360][INFO    ][18125] Executing command ['git', 'rev-parse', '--abbrev-ref', 'HEAD'] in directory '/tmp/foo'
15:31:51,335 [salt.loaded.int.module.cmdmod][ 360][INFO    ][18125] Executing command ['git', 'remote', '--verbose'] in directory '/tmp/foo'
15:31:51,341 [salt.loaded.int.module.cmdmod][ 360][INFO    ][18125] Executing command ['git', 'diff', 'HEAD'] in directory '/tmp/foo'
15:31:51,348 [salt.state][ 232][INFO    ][18125] {'forced update': True}
15:31:51,348 [salt.state][1678][INFO    ][18125] Completed state [/tmp/test-repo] at time 15:31:51.348283
local:
----------
          ID: /tmp/test-repo
    Function: git.latest
      Result: None
     Comment: /tmp/foo is up-to-date, but with local changes. Since 'force_reset' is enabled, these local changes would be reset.
     Started: 15:31:51.297419
    Duration: 50.864 ms
     Changes:
              ----------
              forced update:
                  True

Summary for local
------------
Succeeded: 1 (unchanged=1, changed=1)
Failed:    0
------------
Total states run:     1
Total run time:  50.864 ms
```

## Proceeding with force_reset

```
# salt-call --local state.single git.latest name=/tmp/test-repo target=/tmp/foo force_reset=True
15:32:03,267 [salt.state][ 766][INFO    ][18235] Loading fresh modules for state activity
15:32:03,281 [salt.loaded.int.module.cmdmod][ 360][INFO    ][18235] Executing command ['git', '--version'] in directory '/root'
15:32:03,286 [salt.state][1542][INFO    ][18235] Running state [/tmp/test-repo] at time 15:32:03.285922
15:32:03,286 [salt.state][1564][INFO    ][18235] Executing state git.latest for /tmp/test-repo
15:32:03,287 [salt.loaded.int.states.git][ 558][INFO    ][18235] Checking remote revision for /tmp/test-repo
15:32:03,287 [salt.loaded.int.module.cmdmod][ 360][INFO    ][18235] Executing command ['git', 'ls-remote', '/tmp/test-repo'] in directory '/root'
15:32:03,295 [salt.loaded.int.module.cmdmod][ 360][INFO    ][18235] Executing command ['git', 'for-each-ref', '--format', '%(refname:short)', 'refs/heads/'] in directory '/tmp/foo'
15:32:03,303 [salt.loaded.int.module.cmdmod][ 360][INFO    ][18235] Executing command ['git', 'for-each-ref', '--format', '%(refname:short)', 'refs/tags/'] in directory '/tmp/foo'
15:32:03,310 [salt.loaded.int.states.git][ 112][INFO    ][18235] Checking local revision for /tmp/foo
15:32:03,311 [salt.loaded.int.module.cmdmod][ 360][INFO    ][18235] Executing command ['git', 'rev-parse', 'HEAD'] in directory '/tmp/foo'
15:32:03,320 [salt.loaded.int.states.git][ 121][INFO    ][18235] Checking local branch for /tmp/foo
15:32:03,321 [salt.loaded.int.module.cmdmod][ 360][INFO    ][18235] Executing command ['git', 'rev-parse', '--abbrev-ref', 'HEAD'] in directory '/tmp/foo'
15:32:03,326 [salt.loaded.int.module.cmdmod][ 360][INFO    ][18235] Executing command ['git', 'remote', '--verbose'] in directory '/tmp/foo'
15:32:03,331 [salt.loaded.int.module.cmdmod][ 360][INFO    ][18235] Executing command ['git', 'diff', 'HEAD'] in directory '/tmp/foo'
15:32:03,337 [salt.loaded.int.module.cmdmod][ 360][INFO    ][18235] Executing command ['git', 'rev-parse', '--abbrev-ref', 'master@{upstream}'] in directory '/tmp/foo'
15:32:03,345 [salt.loaded.int.module.cmdmod][ 360][INFO    ][18235] Executing command ['git', 'reset', '--hard', '189ed248555abd3ce9f33318144034d640203773'] in directory '/tmp/foo'
15:32:03,351 [salt.loaded.int.module.cmdmod][ 360][INFO    ][18235] Executing command ['git', 'rev-parse', 'HEAD'] in directory '/tmp/foo'
15:32:03,357 [salt.state][ 232][INFO    ][18235] {'forced update': True}
15:32:03,357 [salt.state][1678][INFO    ][18235] Completed state [/tmp/test-repo] at time 15:32:03.356964
local:
----------
          ID: /tmp/test-repo
    Function: git.latest
      Result: True
     Comment: Repository /tmp/foo is up-to-date

              Changes made: Local changes were discarded. Repository was hard-reset to remote HEAD (189ed24).
     Started: 15:32:03.285922
    Duration: 71.042 ms
     Changes:
              ----------
              forced update:
                  True

Summary for local
------------
Succeeded: 1 (changed=1)
Failed:    0
------------
Total states run:     1
Total run time:  71.042 ms
```

## Local changes are now gone, and state output reflects that when re-run

```
# salt-call --local state.single git.latest name=/tmp/test-repo target=/tmp/foo
15:32:37,453 [salt.state][ 766][INFO    ][18557] Loading fresh modules for state activity
15:32:37,465 [salt.loaded.int.module.cmdmod][ 360][INFO    ][18557] Executing command ['git', '--version'] in directory '/root'
15:32:37,471 [salt.state][1542][INFO    ][18557] Running state [/tmp/test-repo] at time 15:32:37.470513
15:32:37,471 [salt.state][1564][INFO    ][18557] Executing state git.latest for /tmp/test-repo
15:32:37,471 [salt.loaded.int.states.git][ 558][INFO    ][18557] Checking remote revision for /tmp/test-repo
15:32:37,472 [salt.loaded.int.module.cmdmod][ 360][INFO    ][18557] Executing command ['git', 'ls-remote', '/tmp/test-repo'] in directory '/root'
15:32:37,480 [salt.loaded.int.module.cmdmod][ 360][INFO    ][18557] Executing command ['git', 'for-each-ref', '--format', '%(refname:short)', 'refs/heads/'] in directory '/tmp/foo'
15:32:37,487 [salt.loaded.int.module.cmdmod][ 360][INFO    ][18557] Executing command ['git', 'for-each-ref', '--format', '%(refname:short)', 'refs/tags/'] in directory '/tmp/foo'
15:32:37,494 [salt.loaded.int.states.git][ 112][INFO    ][18557] Checking local revision for /tmp/foo
15:32:37,495 [salt.loaded.int.module.cmdmod][ 360][INFO    ][18557] Executing command ['git', 'rev-parse', 'HEAD'] in directory '/tmp/foo'
15:32:37,505 [salt.loaded.int.states.git][ 121][INFO    ][18557] Checking local branch for /tmp/foo
15:32:37,506 [salt.loaded.int.module.cmdmod][ 360][INFO    ][18557] Executing command ['git', 'rev-parse', '--abbrev-ref', 'HEAD'] in directory '/tmp/foo'
15:32:37,512 [salt.loaded.int.module.cmdmod][ 360][INFO    ][18557] Executing command ['git', 'remote', '--verbose'] in directory '/tmp/foo'
15:32:37,518 [salt.loaded.int.module.cmdmod][ 360][INFO    ][18557] Executing command ['git', 'diff', 'HEAD'] in directory '/tmp/foo'
15:32:37,524 [salt.loaded.int.module.cmdmod][ 360][INFO    ][18557] Executing command ['git', 'merge-base', '--is-ancestor', '189ed248555abd3ce9f33318144034d640203773', '189ed248555abd3ce9f33318144034d640203773'] in directory '/tmp/foo'
15:32:37,532 [salt.loaded.int.module.cmdmod][ 360][INFO    ][18557] Executing command ['git', 'rev-parse', '--abbrev-ref', 'master@{upstream}'] in directory '/tmp/foo'
15:32:37,538 [salt.loaded.int.module.cmdmod][ 360][INFO    ][18557] Executing command ['git', 'rev-parse', 'HEAD'] in directory '/tmp/foo'
15:32:37,543 [salt.state][ 232][INFO    ][18557] Repository /tmp/foo is up-to-date
15:32:37,544 [salt.state][1678][INFO    ][18557] Completed state [/tmp/test-repo] at time 15:32:37.543530
local:
----------
          ID: /tmp/test-repo
    Function: git.latest
      Result: True
     Comment: Repository /tmp/foo is up-to-date
     Started: 15:32:37.470513
    Duration: 73.017 ms
     Changes:

Summary for local
------------
Succeeded: 1
Failed:    0
------------
Total states run:     1
Total run time:  73.017 ms
```

**Comment by boltronics** [comment]
Thanks for the quick fix @terminalmage. That's really great to see, and quite a relief. I haven't tested this myself yet, but the pasted results look fine to my eyes, and I've skimmed over the patch. I'll be adding it to my packaging and testing on Monday as a matter of priority.
